### PR TITLE
Settings.toml-configurable pystack

### PR DESCRIPTION
**Pull request by bill88t**
This pr is the replacement for #7396.
It aims to provide a way to configure pystack size after the build.
It also will provide a way to use stackless without rebuilding.

Currently working:

- [x] pystack is now a supervisor allocation.
- [x] Resize via settings.toml

~~Support stackless by setting value to 0.~~

### CODE REVIEW

**Comment by bill88t** [comment]
Stackless will not be implemented with this pr as it would require modification of every gc function call.
And honestly there isn't much benefit to that since adjustments can be made on-the-fly.

Also, the settings.toml variable `CIRCUITPY_PYSTACK_SIZE` will be multiplied by `size_t`, so that the user cannot give invalid sizes (`pystack_size % sizeof(size_t) == 1`)

**Comment by dhalbert** [comment]
> it would require modification of every gc function call

Could you explain why this is? Could you use a global instead?

**Comment by bill88t** [comment]
Sure, `gc_collect_start` initially defined in `py/gc.h`:61 and populated in `py/gc.c`:365 contains the problematic piece of code:
```
#if MICROPY_ENABLE_PYSTACK
// Trace root pointers from the Python stack.
ptrs = (void **)(void *)MP_STATE_THREAD(pystack_start);
gc_collect_root(ptrs, (MP_STATE_THREAD(pystack_cur) - MP_STATE_THREAD(pystack_start)) / sizeof(void *));
#endif
```
This code block is disabled on stackless.
However, leaving it in, when a pystack is not allocated and initialised leaves `pystack_start` and all of those variables undefined.
These variables cannot be reused as placeholders as they would make the system try to use the stack as if it's present there.
And regardless, we would have to get pystack & pystack_size in here forcing the modification of the function definition and the definition of gc_collect in turn. And the definition of gc_collect is used pretty much everywhere.
Making it global would perhaps help, but still we would have to do some weird hacks to force gc to not run that piece of code.
As a reminder, I am still a noob in microcontroller stuff and my C experience is rather small, I may be wrong in all of this.

File: main.c [hunk 1]
```diff
@@ -132,7 +132,7 @@ static void reset_devices(void) {
     #endif
 }
 
-STATIC void start_mp(supervisor_allocation *heap) {
+STATIC void start_mp(supervisor_allocation *heap, supervisor_allocation *pystack, int pystack_size) {
```
Comment (gneverov) [inline main.c:135]: Although passing the stack allocation as a parameter is a sensible thing to do, the trouble here is that that parameter is dependent on MICROPY_ENABLE_PYSTACK. I don't think it is neat to #ifdef the function signature. So I would probably move the heap and stack  allocation calls inside `start_mp` to minimize #ifdefs and duplicated code.

File: main.c [hunk 2]
```diff
@@ -969,6 +978,38 @@ int __attribute__((used)) main(void) {
     alarm_reset();
     #endif
 
+    // Pystack variables have to exist even in stackless for the function calls
```
Comment (gneverov) [inline main.c:981]: getenv.c should be usable from pure C code. I think it only depends on Python for throwing exceptions. The exception handling code can be moved to shared-bindings/os, then shared-module/os/getenv.c can be used from pure C. This avoids the trouble of re-allocating the python stack once python is up.

File: main.c (hunk 1 above)
Comment (bill88t) [reply main.c:135]: In stackless, pystack is being passed uninitilized. The warning is suppressed with a `#pragma`.
The function decleration is not `#ifdef`'ed.
Also, start_mp is being run multiple times. Meanwhile the way I am doing it, the pystack is allocated once.

Comment (bill88t) [reply main.c:135]: Perhaps it would be good to try to have it in `start_mp` regardless, so that it gets resized on reload.

Comment (tannewt) [reply main.c:135]: I do think you want to pass in the pystack allocation because you'll want to free it after mp completes. You don't need to pass in the size though because you can do `get_allocation_length()` like is done for the haep.

File: main.c [hunk 3]
```diff
@@ -730,7 +733,7 @@ STATIC bool run_code_py(safe_mode_t safe_mode, bool *simulate_reset) {
 
 vstr_t *boot_output;
 
-STATIC void __attribute__ ((noinline)) run_boot_py(safe_mode_t safe_mode) {
```
Comment (tannewt) [inline main.c:733]: Why remove this attribute? It doesn't seem related to this change.

File: main.c [hunk 4]
```diff
@@ -969,6 +978,38 @@ int __attribute__((used)) main(void) {
...
+    supervisor_allocation *pystack;
+    mp_int_t pystack_size = CIRCUITPY_PYSTACK_SIZE; // Use build default for now.
+    #if MICROPY_ENABLE_PYSTACK
+    // Allocate default at least temporarily, needed for getenv
+    pystack = allocate_memory(CIRCUITPY_PYSTACK_SIZE, false, false);
+
+    // Fetch value if exists from settings.toml
+    #if CIRCUITPY_OS_GETENV
+    // Init needed.
+    supervisor_allocation *heap = allocate_remaining_memory();
```
Comment (tannewt) [inline main.c:991]: This is extra and done at a deeper level. You'll want to do the pystack deeper as well so that its size can be changed between reloads.

**Review by tannewt — CHANGES_REQUESTED** [review]
Thank you so much for persevering on this! You are getting there. I added a few suggestions around the structure.

**Comment by bill88t** [comment]
Yea I kinda reworked it a bit, now it realloc's on reload (if needed) and it no longer multiplies the value it gets by sizeof(size_t). It got too confusing to calculate, even for me who wrote it.

Ended up turning them back to globals in hopes of simplifying.

**Comment by bill88t** [comment]
Pystack has to be allocated before heap, so this means it will have to be allocated before `start_mp`, hence the function I made.

File: main.c (hunk 3 above)
Comment (bill88t) [reply main.c]: I have put it back.

File: main.c (hunk 4 above)
Comment (bill88t) [reply main.c]: Now all are done right before heap allocations and are free'ed on vm cleanup.

File: main.c (hunk 2 above)
Comment (bill88t) [reply main.c]: Yea it works a-ok without heap. Big oops from me. Fixed it.

**Comment by bill88t** [comment]
Added error messages, and safety for allocation failures. Now it's pretty much perfect.

File: main.c [hunk 5]
```diff
@@ -122,8 +122,8 @@
 uint8_t value_out = 0;
 #endif
 
-#if MICROPY_ENABLE_PYSTACK
-static size_t PLACE_IN_DTCM_BSS(_pystack[CIRCUITPY_PYSTACK_SIZE / sizeof(size_t)]);
+#if MICROPY_ENABLE_PYSTACK && CIRCUITPY_OS_GETENV
```
Comment (gneverov) [inline main.c:125]: With "getenv" being used so far down in the stack, I think we should acknowledge that os_getenv is a mandatory module now.

File: main.c [hunk 6]
```diff
@@ -132,7 +132,29 @@ static void reset_devices(void) {
...
-STATIC void start_mp(supervisor_allocation *heap) {
+#if MICROPY_ENABLE_PYSTACK
+STATIC supervisor_allocation __attribute__ ((noinline)) * alloc_pystack(void) {
+    mp_int_t pystack_size = CIRCUITPY_PYSTACK_SIZE;
+    #if CIRCUITPY_OS_GETENV
+    // Fetch value if exists from settings.toml
+    // Leaves size to build default on any failure
+    (void)common_hal_os_getenv_int("CIRCUITPY_PYSTACK_SIZE", &pystack_size);
+    // Check if value is valid
+    if ((CIRCUITPY_PYSTACK_SIZE != pystack_size) && ((pystack_size < 384) || (pystack_size % sizeof(size_t) != 0))) {
```
Comment (gneverov) [inline main.c:143]: I would write this as:
```
if (getenv_err != GETENV_OK) {
    serial_write_compressed(translate("\nWARNING: Invalid CIRCUITPY_PYSTACK_SIZE, defaulting back to build value.\n\n"));
        pystack_size = CIRCUITPY_PYSTACK_SIZE;
}
pystack_size = MAX(pystack_size, CIRCUITPY_MIN_PYSTACK_SIZE);
```
Because:
1. You need to check for a parsing error in `common_hal_os_getenv_int`.
2. I don't think `pystack` not being mod `sizeof(size_t)` is a real error. Just let integer division round it down.
3. I assume that 384 is the minimum stack size, which should be put in a #define CIRCUITPY_MIN_PYSTACK_SIZE. And again, I don't think being outside this is a error of note.

File: main.c [hunk 7]
```diff
@@ -399,10 +422,11 @@ STATIC bool run_code_py(safe_mode_t safe_mode, bool *simulate_reset) {
...
         #endif
 
+        #if MICROPY_ENABLE_PYSTACK
+        supervisor_allocation *pystack = alloc_pystack();
+        #endif
         supervisor_allocation *heap = allocate_remaining_memory();
-
-        // Prepare the VM state.
-        start_mp(heap);
+        start_mp(heap, pystack);
```
Comment (gneverov) [inline main.c:429]: The problem here is that `pystack` is not defined if not MICROPY_ENABLE_PYSTACK, which is a compile error.

Unfortunately the #define really makes a mess of what is otherwise simple code. Idk, the best solution, but you might consider creating a struct that contains heap and pystack if enabled. Return this struct from `start_mp` and pass it to `stop_mp`. This way all the MICROPY_ENABLE_PYSTACK nonsense is contained to this struct and those functions.

File: main.c [hunk 8]
```diff
@@ -132,7 +132,29 @@ static void reset_devices(void) {
     #endif
 }
 
-STATIC void start_mp(supervisor_allocation *heap) {
+#if MICROPY_ENABLE_PYSTACK
+STATIC supervisor_allocation __attribute__ ((noinline)) * alloc_pystack(void) {
```
Comment (gneverov) [inline main.c:136]: Do you need `noinline? I'm not sure why it is needed on other functions.

File: main.c (hunk 5 above)
Comment (bill88t) [reply main.c:125]: Personally, I feel like super-slim builds still have their place.
getenv isn't needed in those.

File: main.c (hunk 6 above)
Comment (bill88t) [reply main.c:143]: 1. No, if for any reason it fails, it will not update the value. The same thing was done on web workflow (from which I copypasta'ed).
2. Uhh, it sure could be rounded down, as a matter of fact imma do it right now.
3. Yea, a guestimation on my part. Wouldn't call it the ideal value but fine. Someone should test that.

File: main.c (hunk 7 above)
Comment (bill88t) [reply main.c:429]: An excellent idea, doing it.

Comment (bill88t) [reply main.c]: Implemented struct stacks that contains pystack and heap.
If stackless only heap.

File: main.c [hunk 9]
```diff
@@ -122,17 +122,51 @@
...
+} stacks;
+
 static void reset_devices(void) {
     #if CIRCUITPY_BLEIO_HCI
     bleio_reset();
     #endif
 }
 
-STATIC void start_mp(supervisor_allocation *heap) {
+#if MICROPY_ENABLE_PYSTACK
```
Comment (gneverov) [inline main.c:142]: I don't think you meant to have the #if around this function.
You should build this change without MICROPY_ENABLE_PYSTACK to ensure it still builds.

File: main.c [hunk 10]
```diff
@@ -122,17 +122,51 @@
...
+#if MICROPY_ENABLE_PYSTACK && CIRCUITPY_OS_GETENV
+#include "shared-module/os/__init__.h"
 #endif
 
+typedef struct {
+    #if MICROPY_ENABLE_PYSTACK
+    supervisor_allocation *pystack;
+    #endif
+    supervisor_allocation *heap;
+} stacks;
```
Comment (gneverov) [inline main.c:134]: I think we could come up with a better name for this. I suggest:
stacks -> vm_memory_t (the _t suffix is the convention for typedefs in C)
alloc_stacks -> allocate_vm_memory
combo -> vm_memory

File: main.c (hunk 6 above)
Comment (gneverov) [reply main.c]: Re #1: but then the user doesn't get an error message about their settings file being bad

File: main.c (hunk 8 above)
Comment (gneverov) [reply main.c]: If it's not required then remove it.
It tells the compiler not to inline this function, which is something you usually don't care about.

**Review by gneverov — COMMENTED** [review]
It's looking good!

File: main.c (hunk 10 above)
Comment (bill88t) [reply main.c:134]: Yea these names are much better, doing it.

File: main.c (hunk 9 above)
Comment (bill88t) [reply main.c:142]: Fixed the `#if`'s, I will build stackless in a bit.

Comment (bill88t) [reply main.c]: Stackless works just fine. No memory is allocated for pystack.

File: main.c (hunk 8 above)
Comment (bill88t) [reply main.c]: I removed it.

File: main.c (hunk 6 above)
Comment (bill88t) [reply main.c]: Invalid settings.toml is something ignored.
All current getenv code ignores errors (web-workflow & bt-workflow).

File: main.c (hunk 1 above)
Comment (bill88t) [reply main.c]: The `mp_memory_t` struct handles storage and `get_allocation_length()` is used for both pystack and heap now.

**Comment by bill88t** [comment]
Yea this isn't going anywhere, I'll have to remove something to make space.

**Comment by bill88t** [comment]
That code was never updated for pystack so it should have been gone already.

File: main.c (hunk 6 above)
Comment (gneverov) [reply main.c]: Despite precedent, I don't think that is a good practice. If I make a typo in my settings, I'd benefit from a diagnostic telling me so, instead of scratching my head wondering why nothing happened.

**Comment by gneverov** [comment]
Looks good to me! But @tannewt has to approve. Thanks for working on this and persevering making changes.

File: main.c (hunk 6 above)
Comment (dhalbert) [reply main.c]: I think just  "Invalid CIRCUITPY_PYSTACK_SIZE" is probably good enough, and it uses existing strings; I agree ignoring is not so great. It will push the user to documentation.

**Comment by bill88t** [comment]
Alright, I am done with it too.
I tried to further optimise strings, but couldn't.
Should no other changes be requested this PR is final.

File: main.c (hunk 6 above)
Comment (bill88t) [reply main.c]: `CIRCUITPY_PYSTACK_SIZE` has never been used as a string before. It's just a nameshake with the `#if`.
The plan was to get rid of all the compile-time settings for stacks in favor of this implementation.
However they both remained in the end.
I do not see a better name that isn't also used by `#if`'s.
And there is no real risk of users confusing compilation settings with settings.toml variables.
Pretty much nobody will be using stackless after this pr is merged.
There will be no benefit.

I will try to change it to `Invalid %q`.

(I almost missed this discussion comment, temporarily reopening)

Comment (bill88t) [reply main.c]: Yea, `\n`'s are needed so that we are not displaying the warning on a used line. And using multiple writes increases binary size. We have 104 bytes free. I don't wan't to play with core anymore. Me ooga booga :b:ython.

File: main.c [hunk 11]
```diff
@@ -122,17 +122,49 @@
...
+} vm_memory_t;
+
 static void reset_devices(void) {
     #if CIRCUITPY_BLEIO_HCI
     bleio_reset();
     #endif
 }
 
-STATIC void start_mp(supervisor_allocation *heap) {
+STATIC vm_memory_t allocate_vm_memory(void) {
```
Comment (tannewt) [inline main.c:142]: I think you should switch this to `allocate_pystack` and then return the allocation instead of a new type. The heap allocations can then stay where they are.

File: locale/circuitpython.pot [hunk 12]
```diff
@@ -35,6 +35,20 @@ msgid ""
...
+"\n"
+msgstr ""
+
+#: main.c
+msgid ""
+"\n"
+"WARNING: Invalid CIRCUITPY_PYSTACK_SIZE, defaulting back to build value.\n"
+"\n"
+msgstr ""
+
```
Comment (tannewt) [inline locale/circuitpython.pot:51]: Strings take up a lot of space and I don't think these will be hit very often. I'd rather omit them or reuse an existing message.

File: main.c (hunk 7 above)
Comment (tannewt) [reply main.c]: My preference would to be always passing in a pystack and having it NULL when disabled.

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for continuing to revise this. I wonder if we should sit on it until 9.0 because we're removing a function.

File: main.c (hunk 11 above)
Comment (gneverov) [reply main.c:142]: The issue with that approach is that it proliferates `#if MICROPY_ENABLE_PYSTACK` checks throughout the code. That's how we landed on the `vm_memory_t` abstraction.

File: main.c (hunk 7 above)
Comment (bill88t) [reply main.c:435]:
```
#else
pystack = NULL;
```

File: locale/circuitpython.pot (hunk 12 above)
Comment (bill88t) [reply locale/circuitpython.pot]: Did as discussed on discord. [Link](https://discord.com/channels/327254708534116352/327298996332658690/1075883666380701766) to nearby message.

File: main.c (hunk 11 above)
Comment (bill88t) [reply main.c]: Did as requested. Personally I liked the struct more.

**Comment by bill88t** [comment]
The requested changed have been performed. Please review them at your discretion.

Question:
> because we're removing a function.

What function?

File: main.c [hunk 13]
```diff
@@ -399,10 +425,14 @@ STATIC bool run_code_py(safe_mode_t safe_mode, bool *simulate_reset) {
         };
         #endif
 
+        supervisor_allocation *pystack;
```
Comment (gneverov) [inline main.c:428]: You could have
```
supervisor_allocation *pystack = NULL;
```
and avoid the `#else` branch.

Comment (bill88t) [reply main.c]: Thank you very much!

**Comment by bill88t** [comment]
Regarding `supervisor.runtime.next_stack_limit`, it was never updated for pystack. Discord talk [link](https://discord.com/channels/327254708534116352/327298996332658690/1072625346643505162).

If it served some purpose I will gladly put it back, but from what I understood it's an old implementation of this PR basically, configurable stack.

File: main.c (hunk 6 above)
Comment (bill88t) [reply main.c]: This was solved.

**Comment by bill88t** [comment]
Once again, I think I am done.

As a final note, I remind you all that the old `next_stack_limit` was already broken and did not provide any functionality.
Should you guys still want it back in and removed at 9.0, I will gladly put it back, however it will still do nothing.

**Comment by bill88t** [comment]
Turns out `next_stack_limit` has some purpose. I put it back.
If CI doesn't pass I will just disable settable pystack for the failing boards.

**Comment by bill88t** [comment]
Alright and `kicksat-sprite` doesn't need to have `rainbowio` disabled. It just won't have resizable pystack.
So this pr removes no functionallity from any board.

**Comment by dhalbert** [comment]
`kickstat-sprite` has no use for `rainbowio`. The board is designed for space satellites, which don't need NeoPixel animations. If you can enable `pystack` on it with `rainbowio` disabled, that is better, since it's more likely to be useful for how it's used.

**Comment by bill88t** [comment]
This sadly doesn't build. 400 more bytes are needed.

**Comment by dhalbert** [comment]
`_pixelmap` and/or `adafruit_pixelbuf` could be dropped from `kicksat-sprite`. Test a build locally with the largest language that doesn't fit. (`make BOARD=kicksat-sprite TRANSLATION=...`)

**Comment by bill88t** [comment]
```
244240 bytes used, 1520 bytes free in flash firmware space out of 245760 bytes (240.0kB).
31044 bytes used, 165564 bytes free in ram for stack and heap out of 196608 bytes (192.0kB).
```

**Review by tannewt — APPROVED** [review]
Looks good to me! Thank you @bill88t!

**Comment by anecdata** [comment]
Usage is via settings.toml, e.g.: `CIRCUITPY_PYSTACK_SIZE=2048`?

And `supervisor.runtime.next_stack_limit` is intact and adjusts the C stack?

**Comment by bill88t** [comment]
Precisely.
